### ISSUE TITLE
Compiler: add custom byte- and bit- integers

### ISSUE THREAD

**Issue by arekbulski** [kaitai-io/kaitai_struct#313]
After reading #121 and getting totally lost in it, I think that Kaitai should do something similar to Construct BytesInteger and BitsInteger classes. That is, integer types other than basic ones (1-2-4-8 bytes) should be mapped to KaitaiStream method that handles arbitrary sizes, arbitrary endianess:
```
u3  (24-bit)   -->  KaitaiStream._io.read_integer(3, signed=False, swapped=False)
u16 (128-bit)  -->  KaitaiStream._io.read_integer(16, signed=False, swapped=False)
```
Bit- integers are a bit problematic when it comes to little-endianness, but aside of that its also simple.
```
b3  (3-bit)   -->  KaitaiStream._io.read_bits_integer(3, signed=False, swapped=False)
b16 (16-bit)  -->  KaitaiStream._io.read_bits_integer(16, signed=False, swapped=False)
```
I think both methods could be combined into one, but not sure about `align_to_byte()`.

**Comment by KOLANICH** [comment]
see also #76

**Comment by arekbulski** [comment]
#76 is much broader in scope than what I am proposing. Its not entirely related.

**Comment by GreyCat** [comment]
I assume that you'd also want to allow something like calculated parameter to be passed into a type specification? I.e. something like `type: u(x * 2 - 3)`?

It's not hard to implement, but there would be some minor obstacles:

* Many statically typed languages need to know size in compile-time (i.e. all that `short` / `int` / `long` stuff). If we can't derive it, obviously, we'll need to use maximum possible integer.
* Values larger than 64 bits would only fit into some bigint implementations, and there are many platforms that lack standard bigints (i.e. at least C++ and JavaScript).

Existing `read_bits_int()` actually does exactly that: it reads arbitrary number of bits from a stream as an integer. Probably it's not hard to extend `type: b4` parsing to support stuff like `type: b(x * 2 - 3)`.

**Comment by arekbulski** [comment]
Gosh no, I would not expect anything but literals. Its just that it would support things like 24-bit integers without some really weird and strange looking schemas.

**Comment by arekbulski** [comment]
On side note, I think some of the features in Kaitai have hit what is known as analysis paralysis, in particular C runtime. If there is no std bigint in eg. C++, we should just pick a popular one instead of stalemating development.

**Comment by arekbulski** [comment]
I noticed that `read_bits_int()` does not have parameters for signed/unsigned or little/bigendian. Is that function actually emitted by any field?

**Comment by GreyCat** [comment]
> On side note, I think some of the features in Kaitai have hit what is known as analysis paralysis, in particular C runtime.

It's more like a question of priorities. Of course, everything can be solved one way or another, but I don't have time for it right now (actually, the topmost priority for me is to push 0.8 out at last, and then switch to finish general infrastructure improvements). If anyone else would be happy to champion it, I would totally have nothing against it.

> I noticed that `read_bits_int()` does not have parameters for signed/unsigned or little/bigendian. Is that function actually emitted by any field?

Yeah, `type: bX` emits it. It's a long going question with little/big-endian there too - #155.

**Comment by arekbulski** [comment]
I dont think that I have sufficient understanding oh Kaitai to add a C runtime to it. :slightly_frowning_face: I would be willing, just not very able.

**Comment by estan** [comment]
> Gosh no, I would not expect anything but literals. Its just that it would support things like 24-bit integers without some really weird and strange looking schemas.

I on the other hand would think it a really good thing if expressions were allowed. For example the HDF5 format supports different sizes for the offsets/lengths used in the file (official library supports 2, 4, 8 and 16 byte offsets/lengths). The byte size of offsets/lengths are stored in the HDF5 superblock. For such cases it would be great if I could say e.g. `u(superblock.size_of_offsets * 8)`. I'm sure there are other formats where the size of certain fields may vary from file to file and the sizes used are declared somewhere.

I understand sizes > 64 bits may be problematic, but it would be a really good thing to support arbitrary size integers. Or at least 128 bit.

**Comment by GreyCat** [comment]
Choice of standard integer types is perfectly possible using switch-on / cases type. I really doubt that addition of `u(expression)` would provide any additional value here.

128-bit integers are actually pretty tricky too, as lots of languages do not support them, and, at very best, have some built-in support for big integers (and even that is missing in C++, JavaScript, Perl, Lua, etc).

**Comment by estan** [comment]
@GreyCat Yes, perhaps my example was bad. But the point was that the HDF5 format allows any size (2, 4, 8 and 16 is what the official library decided to support), so they could be e.g. 24 bit or 48 bit.

But yes, good to know it'll be easy enough to support at least 2, 4, 8 using switch-on / cases. I think that'll probably cover 99.99% of HDF5 files out there (and I'm willing to bet at least 99% are 8 byte, since the official library will default to using whatever `sizof(long)` is on the machine).

**Comment by estan** [comment]
Also, I should mention I'm just a drive-by commenter that happened to come across this very cool project. So don't put too much weight on what I said. Kaitai looks really cool, and I might use it for a HDF5 format debug tool.
